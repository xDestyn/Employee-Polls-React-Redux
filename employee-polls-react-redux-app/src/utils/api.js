import {
  _getUsers,
  _getQuestions,
  _saveQuestion,
  _saveQuestionAnswer,
  formatQuestion,
} from './_DATA.js';

export const getInitialData = () => {
  return Promise.all([_getUsers(), _getQuestions()]).then(([users, questions]) => ({
    users,
    questions,
  }));
};

export const saveQuestion = (question) => {
  return _saveQuestion(question);
};

export const saveQuestionAnswer = (answer) => {
  return _saveQuestionAnswer(answer);
};

export const formatNewQuestion = (question) => {
  return formatQuestion(question);
};
