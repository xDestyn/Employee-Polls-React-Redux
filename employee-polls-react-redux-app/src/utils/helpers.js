/* eslint-disable no-empty-pattern */

import React from 'react';
import { useLocation, useNavigate, useParams } from 'react-router-dom';

export const formatDate = (timestamp) => {
  const date = new Date(timestamp);
  const time = date.toLocaleTimeString('en-US');
  return time.substr(0, 5) + time.slice(-2) + ' | ' + date.toLocaleDateString();
};

export const formatPoll = (question, author, authedUser) => {
  let hasAnswered = false;
  let isAuthor = false;
  let selectedVote = '';

  // Destructure poll
  const {
    id,
    optionOne: {
      votes: [],
      text: optionOneText,
    },
    optionTwo: {
      votes: [],
      text: optionTwoText,
    },
    timestamp,
  } = question;

  // Verify if a vote has been submitted
  // Extract that vote
  if (
    question.optionOne.votes.includes(authedUser) ||
    question.optionTwo.votes.includes(authedUser)
  ) {
    hasAnswered = true;
    if (question.optionOne.votes.includes(authedUser)) {
      selectedVote = question.optionOne.text;
    } else if (question.optionTwo.votes.includes(authedUser)) {
      selectedVote = question.optionTwo.text;
    }
  }

  // Verify if poll is self made
  if (author === authedUser) {
    isAuthor = true;
  }

  const optionOneVotes = question.optionOne.votes.length;
  const optionTwoVotes = question.optionTwo.votes.length;
  const totalVotes = optionOneVotes + optionTwoVotes;

  const optionOnePercentage = Math.round((optionOneVotes / totalVotes) * 100);
  const optionTwoPercentage = Math.round((optionTwoVotes / totalVotes) * 100);

  // Return Poll Model
  return {
    id,
    timestamp,
    name: author.name,
    avatarURL: author.avatarURL,
    optionOneText,
    optionTwoText,
    hasAnswered,
    selectedVote,
    isAuthor,
    totalVotes,
    optionOnePercentage,
    optionTwoPercentage,
  };
};

export const withRouter = (Component) => {
  const ComponentWithRouterProp = (props) => {
    let location = useLocation();
    let navigate = useNavigate();
    let params = useParams();
    return <Component {...props} router={{ location, navigate, params }} />;
  };

  return ComponentWithRouterProp;
};
