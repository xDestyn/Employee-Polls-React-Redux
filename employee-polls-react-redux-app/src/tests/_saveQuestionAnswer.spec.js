// @ts-nocheck
import { _saveQuestionAnswer } from '../utils/_DATA';

describe('_saveQuestionAnswer', () => {
  it('should return true when correct arguments are received', async () => {
    const authedUser = 'sarahedo';
    const id = '6ni6ok3ym7mf1p33lnez';
    const answer = 'optionOne';

    const response = await _saveQuestionAnswer({ authedUser, id, answer });

    expect(response).toBeTruthy();
  });

  it('should return an error if incorrect arguments are received', async () => {
    const authedUser = 'jsmith';
    const id = undefined;
    const answer = undefined;

    await expect(_saveQuestionAnswer({ authedUser, id, answer })).rejects.toMatchSnapshot();
  });
});
