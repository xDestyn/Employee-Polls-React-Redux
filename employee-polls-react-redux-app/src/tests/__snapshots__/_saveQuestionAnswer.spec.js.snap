// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`_saveQuestionAnswer should return an error if incorrect arguments are received 1`] = `"Please provide authedUser, qid, and answer"`;
